pacsync.8.in
.TH pacsync 8 "Mar 17, 2002" "pacsync #VERSION#" ""
.SH NAME
pacsync \- package update frontend for pacman
.SH SYNOPSIS
\fBpacsync  <operation> [package] [package] ...\fP
.SH DESCRIPTION
\fBpacsync\fP is a \fIpackage update\fP frontend for the
\fIpacman\fP package manager.  It keeps track of the versions
of packages installed on the local system and the versions of
packages on a remote server.  This allows you to issue a single
command and have you system's packages brought into sync.
.SH OPERATIONS
.TP
.B "sync"
Download a new copy of the package list from the server.  This
should be done regularly, usually before an "install", "upgrade"
or "sysupgrade" command.
.TP
.B "install <package>"
Install a new package. \fBpacsync\fP will first look for the package
in the cache directory.  If it is not there, it will attempt to
download the file from the server.  On success, the package will
be installed.  (Note that the downloaded file is also copied to
a cache directory on your filesystem.  You can delete it with
"clean").
.TP
.B "upgrade <package>"
Upgrade a package.  This is the same as \fIinstall\fP, except a package
is upgraded (ie, it is already installed in the system).
.TP
.B "report"
Generate an upgrade report.  This operation will compare all local
package versions to remote versions, and then show you a report of
the differences.
.TP
.B "sysupgrade"
Full system upgrade.  This is the same as report except after confirmation,
\fBpacsync\fP will upgrade all packages that are not up to date.
.TP
.B "clean"
This will remove all downloaded packages from the cache directory.  This
can/should be used regularly to free up diskspace.
.SH CONFIGURATION
Configuration options are stored in /etc/pacsync.conf.  Currently there is only
one directive.
.TP
.B "SYNC_SERVER"
This should be set to the full URL of the download server closest to you.
FTP is preferred, but HTTP is supported also, provided you have wget.
.SH AUTHOR
.nf
Judd Vinet <jvinet@zeroflux.org>
.fi


ChangeLog
VERSION         DESCRIPTION
-----------------------------------------------------------------------------
2.6.3         - A couple memory fixes in the new replaces code
2.6.2         - Fixed a memory cleanup bug
              - Aurelien's patch:
                - bug #159 implemented (for -S and -R)
                - fixed a bug with pacman -Sg (pacman was browsing only one
                  db to get groups)
                - fixed a bug with list_merge()
                - fixed some MLK (in dumppkg() and with "-Qi --orphans")
                - now "pacman -Sg" only displays groups (without content)
                  whereas "pacman -Sg target1 target2" displays groups
                  target1 and target2 with content
2.6.1         - Added http download support (Aurelien Foret)
              - Improved makepkg's --builddeps behaviour when called via
                makeworld
              - makepkg's md5 validation now occurs before source extraction
              - makepkg delays fakeroot entry until after option parsing
              - Fixed an argument-passing bug in fakeroot
              - Modified pacman's behaviour wrt provides -- it now allows
                multiple packages to be installed, even if they provide the
                same thing (they were treated as conflicts before)
2.6           - Added group handling, so one can run 'pacman -S kde' and
                install all files from the KDE group
              - Fixed a duplication bug in cascade package removal
              - Added support for virtual provisions with "provides" tags
              - When conflicts are encountered, pacman now offers the chance
                to remove the conflicting packages (provides or literals)
              - Added support for renamed/combined packages with a "replaces"
                tag
              - Added --nostrip option to makepkg
              - Improved --search to list all packages from all repos when
                a search term is omitted
              - Added logging support through syslog()
              - Added fakeroot support to makepkg (RomanK)
              - Added MD5sum generation/validation to makepkg (RomanK)
              - Fixed a progress bar bug (Aurelien Foret)
              - Sorted makepkg's .FILELISTs (Aurelien Foret)
              - Targets are now re-ordered w.r.t. dependencies when
                using -A/-U
              - Modified --search to work when called as -Sys
              - Modified abs to use ABS_ROOT from /etc/abs/abs.conf (Aurelien)
              - Other bug fixes
2.5.1         - Minor bug fixes
2.5           - Added an URL tag to package info
              - Sped up package load times by about 500% by introducing
                a .FILELIST into the package
              - Renamed the install scriptlet from ._install to .INSTALL
              - Added patch from Aurlien Foret:
                - Better lock handling (RW and RO)
                - Sorted package order in -Qi's dependency lists
                - Added a DBPath option to pacman.conf
                - Fixed memory leaks
                - Added the --nodeps option to -S
2.4.1         - Fixed a bug in makepkg's option parsing
2.4           - Added getopt-style options to makeworld
              - Added -w <destdir> to makepkg
              - makeworld now properly handles packages with --builddeps
              - Added patches from Aurelien Foret:
                - Cascading package removals. -Rc will remove a package and
                  all packages that require it
                - Support for getopt-style options in makepkg
                - the REQUIREDBY field is now updated when a package is
                  removed with --nodeps, then re-installed.
                - Avoids duplicate dependency checks with sync
                - Added a "NoPassiveFtp" option in pacman.conf
                - Improvements to the --query listings
2.3.2         - Added patches from Aurelien Foret:
                - FTP resume feature
								  - removed the hit-^C-and-delete-the-downloading-file safety
                - IgnorePkg option in pacman.conf
                - FTPAGENT setting in makepkg.conf
                - Added --cleancache option to makepkg
2.3.1         - Fixed the progress bar overflow
              - Pacman does not ask "Are you sure" when you use --downloadonly
              - Switched up a couple makepkg options to be more consistent
                with pacman's options
              - If you ^C out of a file download, the package will now be
                removed from the cache directory
2.3           - The beginnings of source-side dependency resolution, makepkg
                can now either A) download/install missing deps with pacman
                -S; or B) find missing deps in the /usr/abs tree and
                build/install them.
              - Added a --nodeps option to makepkg
              - Improved the --search output
2.2           - More bugfixes
              - Added --downloadonly switch to --sync
2.1           - Lots of bugfixes
              - Added support for multiple respositories
              - Improved the config file layout
              - Improved dependency resolution and sorting
2.0           - Added dependency functionality
              - Completely new database format, similar to FreeBSD
              - Better internal data structures
              - Merged pacsync functionality into pacman
              - Now uses libftp for all file retrieval (currently only ftp)
              - Automatic dependency resolution, a la apt-get
              - Moved config stuff from /etc/pacsync.conf to /etc/pacman.conf
              - Much better backup file handling, now uses md5 hashes a la rpm
1.23          - Added install/upgrade/remove scripting control
1.22          - Some manpage typo fixes
              - Added --root switch to pacsync
              - Added --help and ability to specify a PKGBUILD to makepkg

src/pacsync.c
/*
 *  pacsync.c
 * 
 *  Copyright (c) 2002 by Judd Vinet <jvinet@zeroflux.org>
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, 
 *  USA.
 */

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <ftplib.h>
/* pacman */
#include "list.h"
#include "package.h"
#include "db.h"
#include "util.h"
#include "pacsync.h"
#include "pacman.h"

static int log_progress(netbuf *ctl, int xfered, void *arg);
static char sync_fnm[25];
static int offset;

/* pacman options */
extern char *pmo_root;
extern char *pmo_dbpath;
extern unsigned short pmo_nopassiveftp;

/* sync servers */
extern PMList *pmc_syncs;
extern int maxcols;

int sync_synctree()
{
	char ldir[PATH_MAX] = "";
	char path[PATH_MAX];
	mode_t oldmask;
	PMList *files = NULL;
	PMList *i;
	int success = 0;

	for(i = pmc_syncs; i; i = i->next) {
		sync_t *sync = (sync_t*)i->data;		
		snprintf(ldir, PATH_MAX, "%s%s", pmo_root, pmo_dbpath);

		/* build a one-element list */
		snprintf(path, PATH_MAX, "%s.db.tar.gz", sync->treename);
		files = list_add(files, strdup(path));

		success = 1;
		if(downloadfiles(sync->servers, ldir, files)) {
			fprintf(stderr, "failed to synchronize %s\n", sync->treename);
			success = 0;
		}
		list_free(files);
		files = NULL;
		snprintf(path, PATH_MAX, "%s/%s.db.tar.gz", ldir, sync->treename);

		if(success) {
			snprintf(ldir, PATH_MAX, "%s%s/%s", pmo_root, pmo_dbpath, sync->treename);
			/* remove the old dir */
			vprint("removing %s (if it exists)\n", ldir);
			rmrf(ldir);

			/* make the new dir */
			oldmask = umask(0000);
			mkdir(ldir, 0755);
			umask(oldmask);

			/* uncompress the sync database */
			vprint("Unpacking %s...\n", path);
			if(unpack(path, ldir)) {
				return(1);
			}
		}
		/* remove the .tar.gz */
		unlink(path);
	}

	return(!success);
}

int downloadfiles(PMList *servers, char *localpath, PMList *files)
{
	int fsz;
	netbuf *control = NULL;
	PMList *lp;
	int done = 0;
	PMList *complete = NULL;
	PMList *i;

	if(files == NULL) {
		return(0);
	}

	for(i = servers; i && !done; i = i->next) {
		server_t *server = (server_t*)i->data;

		if(!strcmp(server->protocol, "ftp")) {
			FtpInit();
			if(!FtpConnect(server->server, &control)) {
				fprintf(stderr, "error: cannot connect to %s\n", server->server);
				continue;
			}
			if(!FtpLogin("anonymous", "arch@guest", control)) {
				fprintf(stderr, "error: anonymous login failed\n");
				FtpQuit(control);
				continue;
			}	
			if(!FtpChdir(server->path, control)) {
				fprintf(stderr, "error: could not cwd to %s: %s\n", server->path,
					FtpLastResponse(control));
				continue;
			}
			if(!pmo_nopassiveftp) {
				if(!FtpOptions(FTPLIB_CONNMODE, FTPLIB_PASSIVE, control)) {
					fprintf(stderr, "warning: failed to set passive mode\n");
				}
			} else {
				vprint("FTP passive mode not set\n");
			}
		} else if(!strcmp(server->protocol, "http")) {
			if(!HttpConnect(server->server, &control)) {
				fprintf(stderr, "error: cannot connect to %s\n", server->server);
				continue;
			}
		}

		/* set up our progress bar's callback */
		if(strcmp(server->protocol, "file")) {
			FtpOptions(FTPLIB_CALLBACK, (long)log_progress, control);
			FtpOptions(FTPLIB_IDLETIME, (long)1000, control);
			FtpOptions(FTPLIB_CALLBACKARG, (long)&fsz, control);
			FtpOptions(FTPLIB_CALLBACKBYTES, (10*1024), control);
		}

		/* get each file in the list */
		for(lp = files; lp; lp = lp->next) {
			char output[PATH_MAX];
			int j, filedone = 0;
			char *fn = (char*)lp->data;
			struct stat st;

			if(is_in(fn, complete)) {
				continue;
			}

			snprintf(output, PATH_MAX, "%s/%s.part", localpath, fn);
			strncpy(sync_fnm, fn, 24);
			for(j = strlen(sync_fnm); j < 24; j++) {
				sync_fnm[j] = ' ';
			}
			sync_fnm[24] = '\0';
			offset = 0;

			if(!strcmp(server->protocol, "ftp")) {
				if(!FtpSize(fn, &fsz, FTPLIB_IMAGE, control)) {
					fprintf(stderr, "warning: failed to get filesize for %s\n", fn);
				}
				if(!stat(output, &st)) {
					offset = (int)st.st_size;
					if(!FtpRestart(offset, control)) {
						fprintf(stderr, "warning: failed to resume download -- restarting\n");
						/* can't resume: */
						/* unlink the file in order to restart download from scratch */
						unlink(output);
					}
				}
				if(!FtpGet(output, fn, FTPLIB_IMAGE, control)) {
					fprintf(stderr, "\nfailed downloading %s from %s: %s\n",
						fn, server->server, FtpLastResponse(control));
					/* we leave the partially downloaded file in place so it can be resumed later */
				} else {
					filedone = 1;
				}
			} else if(!strcmp(server->protocol, "http")) {
				char src[PATH_MAX];
				if(!stat(output, &st)) {
					/* no resume support yet */
					unlink(output);
				}
				snprintf(src, PATH_MAX, "%s%s", server->path, fn);
				if(!HttpGet(output, src, &fsz, control)) {
					fprintf(stderr, "\nfailed downloading %s from %s: %s\n",
						fn, server->server, FtpLastResponse(control));
					/* no resume support yet */
					 unlink(output); 
				} else {
					filedone = 1;
				}
			} else if(!strcmp(server->protocol, "file")) {
				char src[PATH_MAX];
				snprintf(src, PATH_MAX, "%s%s", server->path, fn);
				vprint("copying %s to %s/%s\n", src, localpath, fn);
				/* local repository, just copy the file */
				if(copyfile(src, output)) {
					fprintf(stderr, "failed copying %s\n", src);
				} else {
					filedone = 1;
				}
			}

			if(filedone) {
					char completefile[PATH_MAX];
					if(!strcmp(server->protocol, "file")) {
						char out[56];
						printf(" %s [", sync_fnm);
						strncpy(out, server->path, 33);
						printf("%s", out);
						for(j = strlen(out); j < maxcols-44; j++) {
							printf(" ");
						}
						fputs("] 100% |   LOCAL\n", stdout);
					} else {
						log_progress(control, fsz-offset, &fsz);
					}
					complete = list_add(complete, fn);
					/* rename "output.part" file to "output" file */
					snprintf(completefile, PATH_MAX, "%s/%s", localpath, fn);
					rename(output, completefile);
			}
			printf("\n");
			fflush(stdout);
		}
		if(list_count(complete) == list_count(files)) {
			done = 1;
		}

		if(!strcmp(server->protocol, "ftp")) {
			FtpQuit(control);
		} else if(!strcmp(server->protocol, "http")) {
			HttpQuit(control);
		}
	}

	return(!done);
}

static int log_progress(netbuf *ctl, int xfered, void *arg)
{
	int fsz = *(int*)arg;
	int pct = ((float)(xfered+offset) / fsz) * 100;
	int i, cur;

	printf(" %s [", sync_fnm);
	cur = (int)((maxcols-44)*pct/100);
	for(i = 0; i < maxcols-44; i++) {
		(i < cur) ? printf("#") : printf(" ");
	}
	printf("] %3d%% | %6dK\r", pct, ((xfered+offset)/1024));
	fflush(stdout);
	return(1);
}

/* Test for existance of a package in a PMList* of syncpkg_t*
 * If found, return a pointer to the respective syncpkg_t*
 */
syncpkg_t* find_pkginsync(char *needle, PMList *haystack)
{
	PMList *i;
	syncpkg_t *sync;
	int found = 0;

	for(i = haystack; i && !found; i = i->next) {
		sync = (syncpkg_t*)i->data;
		if(sync && !strcmp(sync->pkg->name, needle)) {
			found = 1;
		}
	}
	if(!found) {
		sync = NULL;
	}

	return sync;
}

/* vim: set ts=2 sw=2 noet: */
